cassiopeia/solver/NeighborJoiningSolver.py

```diff
@@ -14,15 +15,66 @@
 
 
 class NeighborJoiningSolver(DistanceSolver.DistanceSolver):
+    """
+    Implements the Neighbor-Joining algorithm described by Saitou and Nei (1987)
+    as a derived class of DistanceSolver. This class inherits the generic
+    `solve` method, but implements its own procedure for finding cherries by
+    minimizing the Q-criterion between samples.
+
+    Args:
+        character_matrix: A character matrix of observed character states for
+            all samples.
+        meta_data: Any meta data associated with the samples
+        priors: Prior probabilities of observing a transition from 0 to any 
+            character state
+        dissimilarity_map: A dissimilarity map describing the distances between
+            samples.
+        dissimilarity_function: A function by which to compute the dissimilarity
+            map. Optional if a dissimilarity map is already provided.
+        root_sample: A sample in the character matrix to treat as the root. If
+            not provided, a root of (0,...,0) is added to the character matrix.
+            Throws an error is a root_sample is not provided and no dissimilarity
+            function is provided.
+
+    Attributes:
+        character_matrix: The character matrix describing the samples
+        meta_data: Data table storing meta data for each sample
+        priors: Prior probabilities of character state transitions
+        dissimilarity_map: Dissimilarity map describing distances between
+            samples
+        dissimilarity_function: Function to compute the dissimilarity between
+            samples.
+        root_sample: Sample to treat as a root, an index in the dissimlarity
+            map and character matrix.
+        tree: The tree returned by `self.solve()`. None if `solve` has not been
+            called yet.
+
+    """
+
     def __init__(
         self,
         character_matrix: pd.DataFrame,
         meta_data: Optional[pd.DataFrame] = None,
         priors: Optional[Dict[int, str]] = None,
         dissimilarity_map: Optional[pd.DataFrame] = None,
         dissimilarity_function: Optional[Callable] = None,
+        root_sample: Optional[str] = None,
     ):
 
+        if dissimilarity_function is None and root_sample is None:
+            raise DistanceSolver.DistanceSolveError(
+                "Please specify a root sample or provide a dissimilarity "
+                "function by which to add a root to the dissimilarity map"
+            )
```

Comment:
Is this check necessary? It doesn't seem you're using the dissimilarity function to impute a root, and if the root_sample is None then you add the root.

Reply:
Fixed in the newest commit!